common/transport/amqp/src/amqp.ts

```diff
@@ -706,6 +707,11 @@ export class Amqp {
     connectionParameters.host = parsedUrl.hostname;
     connectionParameters.reconnect = false;
     if (parsedUrl.protocol === 'wss:') {
+      /*Codes_SRS_NODE_COMMON_AMQP_99_043: [The `connect` method shall set the `config.sslOptions.agent` option based on the `amqp.webSocketAgent` object passed in the `options` structure via the `setOptions` function.]*/
+      if (this._options && this._options.amqp && this._options.amqp.webSocketAgent) {
```

Comment:
>this._options.amqp.webSocketAgent [](start = 49, length = 33)

I think this would belong in the config object rather than in a separate _options object here. moving it to the config would have a couple of desirable effects:
- group it with the rest of sslOptions
- remove the need to add a setOptions method to the AmqpBase class

Reply:
@pierreca, the intention here is to keep the API consistent with MQTT transport implementation, which uses [setOptions](https://github.com/Azure/azure-iot-sdk-node/blob/026fa2bcef51f5ae8f8a000fa4cb42038bdf089c/common/transport/mqtt/src/mqtt_base.ts#L206) function, which sets a separate [_options](https://github.com/Azure/azure-iot-sdk-node/blob/026fa2bcef51f5ae8f8a000fa4cb42038bdf089c/common/transport/mqtt/src/mqtt_base.ts#L23) object and takes [MqttTransportOptions](https://github.com/Azure/azure-iot-sdk-node/blob/734bd35f0a48b519ee4de233d1762e7ef2d4521c/device/core/src/interfaces.ts#L79) interface instance as a parameter. There's a similar [HttpTransportOptions](https://github.com/Azure/azure-iot-sdk-node/blob/734bd35f0a48b519ee4de233d1762e7ef2d4521c/device/core/src/interfaces.ts#L62) interface used the same way.

Reply:
I agree with this. this comment complements what i'm suggesting in the main discussion thread but it's probably not very clear so i'll try to rephrase:
- keep the options structure the same, as well as the setOptions call. I agree the external API should be consistent.
- In  `device/transport/amqp/src/amqp.ts`, the caller of this connect function is creating the config object that already has an sslOptions property. Insert the websocketAgent property at this stage. ([https://github.com/Azure/azure-iot-sdk-node/blob/master/device/transport/amqp/src/amqp.ts#L289](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/transport/amqp/src/amqp.ts#L289))
- remove all code in this class, the config.sslOptions object is already being passed to the ws constructor.

i think that would work, wouldn't it? and it would keep all sslOptions together internally, while providing a consistent API at the client level (ie all options object are consistent as you're suggesting)

---
In reply to: [301997067](https://github.com/Azure/azure-iot-sdk-node/pull/579#discussion_r301997067) [](ancestors = 301997067)

Reply:
@pierreca, thank you, I get it now, I am on it.

Reply:
@pierreca, done.